### PR TITLE
Update UPP products for GFS v17

### LINKED ISSUE #1948: GFS Post  PGRB Process on WCOSS2 Takes a longer time
### What is wrong?

The GFS Post PGRB process take more than 5 minutes on WCOSS2. It seems the CFP (MPMD) wasn't launched. 

### What should have happened?

The normal GFS PGRB process is expected within 2 minutes.

### What machines are impacted?

WCOSS2

### Steps to reproduce

checkout the develop branch and run the experiment including gfs post process for generating pgrb2/pgrb2b datasets.  

### Additional information

Check runtime log gfs_post*

### Do you have a proposed solution?

Add a line "launcher=$APRUN_DWN" before "chmod 755 ""${mpmd_cmdfile}" in ush/run_mpmd.sh

APRUN_DWN="mpiexec -np 24 -cpu-bind verbose,core cfp"

### LINKED ISSUE #1965: Update g2tmpl from 1.0.10 to 1.0.12
### What is wrong?

Some new GFS V17 products can't be generated with g2tmpl/1.0.10 in UPP.

### What should have happened?

The land fraction should be generated in the GFS master file as:
844:2252411138:d=2019083000:var discipline=2 center=7 local_table=1 parmcat=0 parm=233:surface:6 hour fcst:

### What machines are impacted?

All or N/A

### Steps to reproduce

Check the GFS V17 master file.

### Additional information

None.

### Do you have a proposed solution?

Update g2tmpl to 1.0.12.

### LINKED ISSUE #1994: Add new variables in GFS V17 pgrb2 products
### What new functionality do you need?

Some new variables are needed in pgrb2 dataset for GFS V17 development.

### What are the requirements for the new functionality?

- Turn on merra2 aerosol field generation in the UPP process.
- Add new variables in pbrb2 generation.

### Acceptance Criteria

The following fields are available at pgrb2 dataset:

- 6 AOTK
- HLCY: 1000-0 m above ground
- 2 TSNOWP
- 2 FROZR
- 2 FRZR
- SDEN

### Suggest a solution (optional)

_No response_

### PR DESCRIPTION
**Pull request by WenMeng-NOAA**
# Description
Updates of generating GFS V17 UPP products to support HR3:

- Update g2tmpl to 1.0.12
- Add changes to launch cfp on WCOSS2 to generate pgrb2/pgrb2b files
- Changes for generating new products (MERRA2, six winter weather variables, 1km HLCY, SDEN) in pgrb2 files

Resolves #1965
Resolves #1994

# Type of change

- Bug fix (fixes something broken)
- New feature (adds functionality)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? YES
- Does this change require a documentation update? NO

# How has this been tested?

# Checklist
- [ ] Any dependent changes have been merged and published
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] I have made corresponding changes to the documentation if necessary

### CODE REVIEW

**Comment by WalterKolczynski-NOAA** [comment]
Please fill in the PR template.

**Review by aerorahul — APPROVED** [review]
Changes look good.
Please fill out the PR template.
I assume the change to `POSTGPVARS` is only effective when doing offline post.  
Offline post is currently only exercised in the analysis and `GOES=YES` cases.

File: ush/run_mpmd.sh
Comment (WenMeng-NOAA) [reply ush/run_mpmd.sh]: Done

**Review by WalterKolczynski-NOAA — APPROVED** [review]
Conditionally approved pending successful CI testing.

File: ush/run_mpmd.sh
Comment (aerorahul) [inline ush/run_mpmd.sh]: run_mpmd.sh is generic. 
APRUN_DWN is specific. 
Thanks for catching this, I am not sure how this escaped on WCOSS2 testing of CFP. 
We will need this to be corrected, but we can do that in another PR after this one is merged.

**Comment by WalterKolczynski-NOAA** [comment]
`g2tmpl/1.12.0` isn't installed under `intel/2018.4`, at least on Orion (I didn't check Hera). This may have to wait until other efforts to increase the compiler version are complete.

**Comment by WenMeng-NOAA** [comment]
> `g2tmpl/1.12.0` isn't installed under `intel/2018.4`, at least on Orion (I didn't check Hera). This may have to wait until other efforts to increase the compiler version are complete.

@WalterKolczynski-NOAA It's g2tmpl/1.10.2.

**Comment by aerorahul** [comment]
@WenMeng-NOAA 
Would you be willing to test this patch for the changes in `ush/run_mpmd.sh` on WCOSS2?  I believe this resolve #1948
Thank you.
[mpmd_wcoss2.diff.txt](https://github.com/NOAA-EMC/global-workflow/files/13166547/mpmd_wcoss2.diff.txt)
The changes in the file are:
```diff
diff --git i/ush/run_mpmd.sh w/ush/run_mpmd.sh
index 352a41131..87ed8f80d 100755
--- i/ush/run_mpmd.sh
+++ w/ush/run_mpmd.sh
@@ -4,6 +4,9 @@ source "${HOMEgfs}/ush/preamble.sh"

 cmdfile=${1:?"run_mpmd requires an input file containing commands to execute in MPMD mode"}

+# Determine the number of MPMD processes from incoming ${cmdfile}
+nprocs=$(wc -l < "${cmdfile}")
+
 # Local MPMD file containing instructions to run in CFP
 mpmd_cmdfile="${DATA:-}/mpmd_cmdfile"
 if [[ -s "${mpmd_cmdfile}" ]]; then rm -f "${mpmd_cmdfile}"; fi
@@ -19,7 +22,6 @@ if [[ "${launcher:-}" =~ ^srun.* ]]; then  #  srun-based system e.g. Hera, Orion
     ((nm=nm+1))
   done < "${cmdfile}"

-  nprocs=$(wc -l < "${mpmd_cmdfile}")
   set +e
   # shellcheck disable=SC2086
   ${launcher:-} ${mpmd_opt:-} -n ${nprocs} "${mpmd_cmdfile}"
@@ -42,7 +44,8 @@ elif [[ "${launcher:-}" =~ ^mpiexec.* ]]; then  # mpiexec
   done < "${cmdfile}"

   chmod 755 "${mpmd_cmdfile}"
-  ${launcher:-} "${mpmd_cmdfile}"
+  # shellcheck disable=SC2086
+  ${launcher:-} -np ${nprocs} ${mpmd_out:-} "${mpmd_cmdfile}"
   rc=$?
   if (( rc == 0 )); then
     out_files=$(find . -name 'mpmd.*.out')
```

File: ush/run_mpmd.sh
Comment (WenMeng-NOAA) [reply ush/run_mpmd.sh]: @aerorahul This line is my workaround. 
APRUN_DWN='mpiexec -l -np 24 --cpu-bind verbose,core cfp'
With this setting, the MPMD runtime is within 1.5 minutes.
I saw APRUN_DWN setting from WCOSS2.env in the G-W version for HR2, but not in current develop branch. BTW, I will remove my change from this PR. We can work this later.

Comment (aerorahul) [reply ush/run_mpmd.sh]: Ok. Understood.
Would you be willing to test the patch in https://github.com/NOAA-EMC/global-workflow/pull/1973#issuecomment-1779290357 in lieu of your workaround?
I am ok with including it in this PR or promoting this to a hotfix and getting it in sooner.

**Comment by WenMeng-NOAA** [comment]
@aerorahul Please let me know your file location. I can pick up and test.

**Comment by aerorahul** [comment]
> @aerorahul Please let me know your file location. I can pick up and test.

@WenMeng-NOAA 
I have pushed the changes to [this](https://github.com/aerorahul/global-workflow/tree/hotfix/mpmd_wcoss2) branch.

**Comment by WenMeng-NOAA** [comment]
> > @aerorahul Please let me know your file location. I can pick up and test.
> 
> @WenMeng-NOAA I have pushed the changes to [this](https://github.com/aerorahul/global-workflow/tree/hotfix/mpmd_wcoss2) branch.

@aerorahul Got it. Thanks!

**Comment by WalterKolczynski-NOAA** [comment]
> > `g2tmpl/1.12.0` isn't installed under `intel/2018.4`, at least on Orion (I didn't check Hera). This may have to wait until other efforts to increase the compiler version are complete.
> 
> @WalterKolczynski-NOAA It's g2tmpl/1.10.2.

Orion only has `g2tmpl/1.9.1` and `g2tmpl/1.10.0` installed under intel 2018.4.

**Comment by WalterKolczynski-NOAA** [comment]
Since this did pass Hera, we might need to have them install `g2tmpl/1.10.2` since the stacks should match on both machines.

**Comment by WenMeng-NOAA** [comment]
> > > @aerorahul Please let me know your file location. I can pick up and test.
> > 
> > 
> > @WenMeng-NOAA I have pushed the changes to [this](https://github.com/aerorahul/global-workflow/tree/hotfix/mpmd_wcoss2) branch.
> 
> @aerorahul Got it. Thanks!

@aerorahul Your patching is working in my test. The runtime of mpmd process is within 1.3 minutes. You may look into my runtime log /u/wen.meng/noscrub/ncep_post/gw_v17/driver/gfsv17_post.o107308975-rahul:
```
Begin fv3gfs_downstream_nems.sh at Wed Oct 25 19:00:48 UTC 2023
+ run_mpmd.sh[48](/u/wen.meng/ptmp/gfsv17_test/rundir/poescript): mpiexec -l -np 24 --cpu-bind verbose,core cfp /u/wen.meng/ptmp/gfsv17_test/rundir/mpmd_cmdfile
End fv3gfs_downstream_nems.sh at 19:02:09 with error code 0 (time elapsed: 00:01:21)

```

**Comment by WenMeng-NOAA** [comment]
> > > `g2tmpl/1.12.0` isn't installed under `intel/2018.4`, at least on Orion (I didn't check Hera). This may have to wait until other efforts to increase the compiler version are complete.
> > 
> > 
> > @WalterKolczynski-NOAA It's g2tmpl/1.10.2.
> 
> Orion only has `g2tmpl/1.9.1` and `g2tmpl/1.10.0` installed under intel 2018.4.

@WalterKolczynski-NOAA I will revert the changes for Orion. You may add them when the global-workflow switching to modulefiles  the spack-stack.

**Comment by WalterKolczynski-NOAA** [comment]
Are updates needed now that #1977 has been merged?

**Comment by WenMeng-NOAA** [comment]
> Are updates needed now that #1977 has been merged?

@WalterKolczynski-NOAA Just synced.

**Comment by TerrenceMcGuinness-NOAA** [comment]
**Slurm** canceled jobs on Orion for this PR yesterday because of a maintenance shutdown.  Simply restarting the CI Test on Orion for this PR.

**Comment by WalterKolczynski-NOAA** [comment]
Labeled this as "on hold" until the independent issues on Orion around the analysis are resolved.

**Comment by WalterKolczynski-NOAA** [comment]
@WenMeng-NOAA please pull in the changes from develop and we'll test again.

**Comment by WenMeng-NOAA** [comment]
> @WenMeng-NOAA please pull in the changes from develop and we'll test again.

@WalterKolczynski-NOAA Done.

**Comment by WenMeng-NOAA** [comment]
@WalterKolczynski-NOAA I just synced my branch with develop branch from upstream.

**Pull request merged**
